### PR TITLE
added custom watermark for kinesis reader

### PR DESCRIPTION
**Pull request by sjmittal**
The PR is for attempted fix of the issue #28760 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by damccorm** [comment]
@sjmittal what are next steps here? It looks like currently this PR is not compiling

**Comment by damccorm** [comment]
@damondouglas @Abacn @ahmedabu98 @johnjcasey could someone please take a look at this one?

**Comment by johnjcasey** [comment]
Ack, taking a look

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/EFOShardSubscribersPool.java [hunk 1]
```diff
@@ -331,6 +331,15 @@ Instant getWatermark() {
     return minTimestamp(state.values().stream().map(ShardState::getWatermark));
   }
 
+  Instant getEventTimestamp(KinesisRecord record) {
+    List<ShardState> sharedStates = new ArrayList<>(state.values());
+
+    if (sharedStates.isEmpty()) {
+      return null;
+    }
+    return sharedStates.get(0).getEventTime(record);
```
Comment (johnjcasey) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/EFOShardSubscribersPool.java:340]: Is it valid to get the event time from any shard state?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/ShardReadersPool.java [hunk 2]
```diff
@@ -255,6 +256,16 @@ Instant getLatestRecordTimestamp() {
...
   }
 
+  Instant getEventTimestamp(KinesisRecord record) {
+    ImmutableList<ShardRecordsIterator> shardRecordsIterators =
+        shardIteratorsMap.get().values().asList();
+
+    if (shardRecordsIterators.isEmpty()) {
+      return null;
+    }
+    return shardRecordsIterators.get(0).getEventTime(record);
```
Comment (johnjcasey) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/ShardReadersPool.java:266]: Same question here, is it valid to ask for the event time from any shard record iterator?

**Review by johnjcasey — CHANGES_REQUESTED** [review]
We need to add tests to this change

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/EFOShardSubscribersPool.java (hunk 1 above)
Comment (sjmittal) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/EFOShardSubscribersPool.java:340]: We get watermark also from `ShardState#getWatermark` so it seems logical that event timestamp is also fetched from `ShardState#getEventTime`.

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/ShardReadersPool.java (hunk 2 above)
Comment (sjmittal) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/ShardReadersPool.java:266]: We get watermark also from `ShardRecordsIterator#getWatermark` so it seems logical that event timestamp is also fetched from `ShardRecordsIterator#getEventTime` which internally returns `watermarkPolicy.getTimestamp`.

Comment (johnjcasey) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/ShardReadersPool.java:266]: SGTM

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/EFOShardSubscribersPool.java (hunk 1 above)
Comment (johnjcasey) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/EFOShardSubscribersPool.java:340]: SGTM

**Comment by sjmittal** [comment]
> We need to add tests to this change

I have updated the test cases to `org.apache.beam.sdk.io.aws2.kinesis.WatermarkPolicyTest`.
The essence of this PR is to get the current timestamp of the stream from the record if custom watermark policy is set for it, otherwise if falls back to the arrival time policy.
Hence I have updated the test cases for above class. 

In other test classes where we test the Kinesis reader or shard subscribers the setup itself of these test cases assumes arrival time policy hence there are no changes to these test classes.

**Comment by damccorm** [comment]
@johnjcasey could you please take another look at this one/help take this pr to completion?

**Comment by damccorm** [comment]
@johnjcasey could you please take another look here?

**Pull request closed without merging**